web_sdk/sdk_rewriter.dart

```diff
@@ -82,56 +88,93 @@ void main(List<String> arguments) {
   final Directory directory = Directory(results['output-dir'] as String);
   final String inputDirectoryPath = results['input-dir'] as String;
-  for (final String inputFilePath in results['input'] as Iterable<String>) {
-    final File inputFile = File(inputFilePath);
-    final File outputFile = File(path.join(
-        directory.path, inputFile.path.substring(inputDirectoryPath.length)))
-      ..createSync(recursive: true);
-    final String source = inputFile.readAsStringSync();
-    final String rewrittenContent = rewriteFile(
-      source,
-      filePath: inputFilePath,
-      isUi: results['ui'] as bool,
-      isEngine: results['engine'] as bool,
-    );
-    outputFile.writeAsStringSync(rewrittenContent);
-    if (results['stamp'] != null) {
-      File(results['stamp'] as String).writeAsStringSync('stamp');
+
+  String Function(String source)? preprocessor;
+  List<Replacer> replacementPatterns;
+  String? libraryName;
+  if (results['ui'] as bool) {
+    replacementPatterns = uiPatterns;
+  } else {
+    libraryName = results['library-name'] as String;
+    if (libraryName == null) {
+      throw Exception('library-name must be specified if not rewriting ui');
     }
+    preprocessor = (String source) => preprocessPartFile(source, libraryName!);
+    replacementPatterns = generatePartsPatterns(libraryName);
   }
-}
+  for (final String inputFilePath in results['source-file'] as Iterable<String>) {
+    String pathSuffix = inputFilePath.substring(inputDirectoryPath.length);
+    if (libraryName != null) {
+      pathSuffix = path.join(libraryName, pathSuffix);
+    }
+    final String outputFilePath = path.join(directory.path, pathSuffix);
+    processFile(inputFilePath, outputFilePath, preprocessor, replacementPatterns);
+  }
+
+  if (results['api-file'] != null) {
+    if (libraryName == null) {
+      throw Exception('library-name must be specified if api-file is specified');
+    }
 
-String rewriteFile(String source, {required String filePath, required bool isUi, required bool isEngine}) {
-  final List<Replacer> replacementPatterns = <Replacer>[];
-  replacementPatterns.addAll(sharedPatterns);
-  if (isUi) {
-    replacementPatterns.addAll(uiPatterns);
-  } else if (isEngine) {
-    if (filePath.endsWith('lib/src/engine.dart')) {
-      _validateEngineSource(filePath, source);
-      replacementPatterns.addAll(engineLibraryPatterns);
-    } else {
-      source = _preprocessEnginePartFile(source);
-      replacementPatterns.addAll(enginePartsPatterns);
+    final String inputFilePath = results['api-file'] as String;
+    final String outputFilePath = path.join(
+        directory.path, path.basename(inputFilePath));
+    String extraImports = '';
+    if (libraryName == 'engine') {
+      extraImports = "import 'dart:_skwasm_stub' if (dart.library.ffi) 'dart:_skwasm_impl';\n";
+    } else if (libraryName == 'skwasm_stub' || libraryName == 'skwasm_impl') {
+      extraImports = "import 'dart:_engine';\n";
```

Comment:
Rather than hard-coding extra imports into the GN rules and here, what if we specify them in the source code inside special comments? Such as:

```dart
// ON_REWRITE_ADD: import 'dart:_engine';

// ON_REWRITE_REMOVE
import 'src/engine.dart';
```

Then all the rewriter needs to do is uncomment these lines. This same syntax could be used for other things that we want included in the rewritten version that shouldn't be part of the package version.

Reply:
I actually prefer to reduce the amount of things that the developer needs to do with regards to making the source code work as a normal package vs sdk library. They should write code as if they were writing a normal library, while the rewriter takes care of the rest.

Obviously, we are not 100% there yet. The developer already thinks about the rewriter e.g. when they use imports, they have to do `import 'dart:math' as math;`, they can't do `import 'dart:math';`.

But I like to keep it to a minimum. And in the future, I'm hoping we can get rid of the rewriter altogether without changing anything in the source code (once `dart:ui` becomes a normal package instead of an sdk library).

Reply:
To be honest, this area  has pointed out is the part of this PR that made me feel the most "blech". It is kind of gross, not gonna lie. One thing I had contemplated was to actually collect all the imports as I processed the part files and dynamically generate the list of imports based on that. Only problem is, that would work for every case *except* this actual ones I'm special casing here, so it doesn't really solve my problem in a good way. That is, it works for all the `dart:math` style imports where they are the same before and after the rewriter, but not for internal dart libraries that are created at the time of rewrite.

I also considered just having explicit flags for each sublibrary with the command line tool (like `--ui`, `--engine`, `--skwasm_stub` and `--skwasm_renderer`) so that the special casing is a little more explicit. But this felt a little like six of one half dozen of the other, the responsibilities are the same and it's still special casing so I was like meh.

I hadn't considered Yegor's idea of dropping comments in the dart file, which is interesting. It does shift the responsibility of some of this stuff back out to the creator of the package, but on the other hand it would make it more generically useful (you could add another sublibrary without changing the script). But also, it's a whole other generic preprocessing paradigm that feels a little overengineered for something that is a pretty niche one-off rewriter script.

At the end of the day, after all this hemming and hawing about this stuff, I just figured every one of these has some sort of tradeoff and is imperfect. I don't want "perfect" to be the enemy of "better", and this is already quite a bit more generic and flexible than it was, so I just called it here, even though this special casing is a bit of a hack. This also seems like whatever we land on, it's simple to come back and change if it ever gives us any trouble, so I'm not that concerned about the particular strategy we land on. And, as Mouad mentions, I hope that this rewriter script is not really a permanent installment anyway, and that at some point the whole thing becomes moot in the fullness of time.

If anyone feels really strongly about going down a different path on it I'm happy to change it, but again I don't personally think this is a critical decision either way.

Reply:
I agree that the comment-based extension on what otherwise is just normal Dart language would be a quirk. However, having an sdk_rewriter is also a quirk. My motivation is mostly to reduce the amount of magic that the sdk_rewriter contains. This way each quirk is close to the code that needs it. Currently, the code that needs a quirk has no indication about it, and the implementation of the quirk is somewhere else (the sdk_rewriter). If you refactor that code (rename, move, etc) you will see a weird breakage elsewhere.

Having said that, I don't feel strongly about it. It's a one-off thing limited to the web engine. We deal with the sdk_rewriter once or twice a year, so whichever way we go, it probably won't make a lot of difference.